app/vmalert/manager.go

```diff
@@ -156,11 +156,14 @@ func (m *manager) update(ctx context.Context, groupsCfg []config.Group, restore
 		var wg sync.WaitGroup
 		for _, item := range toUpdate {
 			wg.Add(1)
+			// cache the cancel func before updating the Group, as it might change during the Update
+			evalCancel := item.old.EvalCancelFn
```

Comment:
Hm, should it be `cancel := item.old.EvalCancelFn()`?

Since, direct access to the EvalCancelFn of modified rule may not be safe without mutex protection.  Also, result of method `EvalCancelFn()` never called at this code (maybe it should be `EvalCancelFn()()`.

Reply:
Also, I think, it also possible to collect cancel funcs and call it one by one, after wg.Wait() call.

 It should mitigate possible issue, when in-flight requests for old.Rule will be canceled and new rule didn't start yet. (e.g. rolling update).

Reply:
Thanks! Indeed, I missed the function call. However, after reading [comment](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/5698#discussion_r1467979688) from @Haleygo I think it can be fixed in much simpler way. wdyt?